docs/OtherConfig.md
# Other Configuration Items

The main README.md details the basic config items, however more are supported,
and these are detailed below:


Bliss
-----

```
{
 "bliss":{
  "enabled":true,
  "analyser":"linux/x86-64/bliss-analyse"
 }
}
```

* `bliss.enabled` should be set to true if tracks should be analysed with bliss.
This defaults to true for Linux and macOS when analysing, and to true for API if
tracks have been analysed with Musly.
* `bliss.analyser` should contain the path to the Bliss analyser extractor -path
is relative to `music-similarity.py` Only required if analyising tracks. By
default music-similarity will attempt to set this automatically.


Musly
-----
```
{
 "musly":{
  "enabled":true,
  "lib":"lib/x86-64/fedora/libmusly.so",
  "styletracks":1000,
  "styletracksmethod":"genres",
  "extractlen":120,
  "extractstart":-210
 }
}
```
* `musly.enabled` should be set to true if Musly is to be used for similarity.
This defaults to true for Windows when analysing, and to true for API if tracks
have been analysed with Musly.
* `musly.lib` should contain the path the Musly shared library - path is
relative to `music-similarity.py`
* `musly.styletracks` A  subset of tracks is passed to Musly's `setmusicstyle`
function, by default 1000 random tracks is chosen. This config item can be used
to alter this. Note, however, the larger the number here the longer it takes to
for this call to complete. As a rough guide it takes ~1min per 1000 tracks.
If you change this config item after the jukebox is written you will need to
delete the jukebox file and restart the server. Only used if analyising tracks.
* `musly.styletracksmethod` configures how tracks are chosen for styletracks. If
set to `genres` (which is the default if not set) then the meta-data db is
queried for how many tracks each genre has and tracks are chosen for each of
these genres based upon the percentage of tracks in a genre. If set to `albums`
then at least one track from each album is used. If set to anything else then
random tracks are chosen. Only used if analyising tracks.
* `musly.extractlen` The maximum length in seconds of the track to decode. If
zero or greater than the track's length, then the whole track will be decoded.
Note, however, that only a maximum of 5 minutes is used for analysis. Only used
if analysing tracks.
* `musly.extractstart` The starting position in seconds of the excerpt to
decode. If zero, decoding starts at the beginning. If negative, the excerpt is
centred in the track, but starts at `-extractstart` the latest. If positive and
`extractstart`+`extractlen` exceeds the track length, then the excerpt is taken
from the end of the track. Only used if analyising tracks.


Essentia
--------

Apart from `enabled`, `extractor`, and `highlevel` the following config items
are usually supplied via the LMS plugin when it asks for mixes.


```
{
 "essentia":{
  "enabled":true,
  "extractor":"linux/x86-64/essentia_streaming_extractor_music",
  "highlevel":false,
  "filterattrib_count":4,
  "filterattrib_lim":0.2,
  "filterattrib_cand":0.4,
  "attrmix_yes":0.6,
  "attrmix_no":0.4
 },
 "paths":{
  "cache":"/path/to/store/cache"
 }
}
```

* `essentia.enabled` should be set to true if Essentia is to be used for
filtering, or similairty. This defaults to true if when analysing if the 
extractor is found, and to true for API if tracks have been analysed with
Essentia.
* `essentia.extractor` should contain the path to the Essentia extractor - path
is relative to `music-similarity.py` Only required if analyising tracks. By
default music-similarity will attempt to set this automatically.
* `essentia.highlevel` Specify whether to support high-level Essentia analysis
features. This defaults to true for Linux only.
* `essentia.filterattrib_count` Configures nubmer of attributes to use.
* `essentia.filterattrib_lim` If filtering on attributes, only attributes that
are less than this, or higher than 1.0-this will be used.
* `essentia.filterattrib_cand` When filtering on attributes, the selected
attributes of the seed need to be less than this, or higher than 1.0-this.
* `essentia.attrmix_yes` When using 'Smart Mixes', 'Yes' attributes need to be
higher than or equal to this.
* `essentia.attrmix_no` When using 'Smart Mixes', 'No' attributes need to be
lower than or equal to this.
* `paths.cache` if set, then the full output of Essentia analysis for each track
will be stored within a GZip compressed JSON file (`<music file>.json.gz`)


CUE Tracks
----------

During analysis, this script will also analyse individual CUE tracks. To do this
it needs access to the LMS database file to know the position of each track,
etc. as well as the path to your music files as LMS sees them.

```
{
 "paths":{
  "lms":"/media/Music/",
  "tmp":"/tmp/"
 },
 "lmsdb":"/path/to/lms/Cache/library.db"
}
```

* `paths.lms` should be the path where LMS accesses your music files.
* `lmsdb` should hold the path to the LMS database file. This is only required
for analysis, and only if you have CUE files. `ffmpeg` is required to split
tracks.
* `paths.tmp` When analysing music, this script will create a temporary folder
to hold any split CUE file tracks. The path set here needs to be writable. If
left unset, the default, then it will be set to the system's default temporary
folder path. This config item is only used for analysis.


Normalizing Strings
-------------------

```
{
 "normalize":{
  "artist":["feet", "ft", "featuring"],
  "album":["deluxe edition", "remastered"],
  "title"["demo", "radio edit"]
 }
}
```

* `normalize.artist` List of strings to split artist names, e.g. "A ft. B"
becomes "A" (periods are automatically removed). This is then used to aid
filtering of tracks - i.e. to prevent artists from being repeated in a mix.
* `normalize.album` List of strings to remove from album names. This is then
used to aid filtering of tracks - i.e. to prevent albums from being repeated in
a mix.
* `normalize.title` List of strings to remove from titles. This is then
used to aid filtering of tracks - i.e. to prevent duplicate tracks in the mix.


Genre grouping
--------------

Genres can configured via the `genres` section of `config.json`, using the
following syntax:

```
{
 "genres:[
  [ "Rock", "Hard Rock", "Metal"],
  [ "Pop", "Dance", "R&B"]
 ]
}
```

If a seed track has `Hard Rock` as its genre, then only tracks with `Rock`,
`Hard Rock`, or `Metal` will be allowed. If a seed track has a genre that is not
listed here then any track returned by Musly, that does not contain any genre
listed here, will be considered acceptable. Therefore, if seed is `Pop` then
a `Hard Rock` track would not be considered.

*NOTE* Genre groups can now be configured directly in the [LMS Music Similarity Plugin](https://github.com/CDrummond/lms-musicsimilarity),
therefore there is no need to add these to `config.json`, and this section
is no longer required.


Misc
----

```
{
 "port":11000,
 "host":"0.0.0.0",
 "threads":8,
 "minduration":30,
 "maxduration":1800,
 "excludegenres":[
   "Podcast", "Audiobook"
 ],
 "simalgo":"musly",
 "mixed":{
  "essentia":33,
  "bliss":33,
  "musly:34
 }
}
```

* `port` This is the port number the similarity server is accessible on.
* `host` IP address on which the similarity server will listen on. Use `0.0.0.0`
to listen on all interfaces on your network.
* `threads` Number of threads to use during analysis phase. This controls how
many calls to `ffmpeg` are made concurrently, and how many concurrent tracks
Musly and Essentia are asked to analyse. Defaults to CPU count, if not set.
* `minduration` Only analyse tracks with duration >= this.
* `maxduration` Only analyse tracks with duration <= this.
* `excludegenres` List of genres that should be excluded from analysis. Any
tracks that have a genre from this list will not be analysed.
* `simalgo` Which method to use for similarity score; `musly`, `essentia`,
`bliss`, `mixed`, or `simplemixed`. This defaults to `bliss` for Linux and macOS,
and `musly` for Windows. This only affects the API usage, analysis will use all
enabled types. `mixed` uses a KDTree of all similarity scores, to attempt to
locate the track that is most similar across all algorithms. `simplemixed` is
a simplified version of `mixed` where the average of similarity scores is
used for each track. **NOTE** Both `mixed` and `simplemixed` will be slower, as
these calculate similarites against all tracks, combine, and then produce a
result. How effective, or usefull, this is remains to be seen, and these _might_
be removed.
* `mixed.essentia`, `mixed.bliss`, `mixed.musly` are used to define the
percentage each of these in the similarity score. Only used if `simalgo` is set
to `mixed` or `simplemixed`.


ChangeLog
0.0.1
-----
1. Initial release.


lib/config.py
#
# Analyse files with Musly, Essentia, and Bliss, and provide an API to retrieve similar tracks
#
# Copyright (c) 2021-2022 Craig Drummond <craig.p.drummond@gmail.com>
# GPLv3 license.
#

import json, logging, os, pathlib, platform
from . import tracks_db

_LOGGER = logging.getLogger(__name__)
# Only Linux and mac, for now, support highlevel analysis
SUPPORT_ESSENTIA_HIGHLEVEL = ['linux', 'mac']
# Only Linux, for now, supports bliss
SUPPORT_BLISS = ['linux', 'mac']

def exit_with_error(s):
    _LOGGER.error(s)
    exit(-1)


def fix_path(path):
    '''
    Replace env vars in paths
    '''
    for e  in ['HOME', 'USERPROFILE', 'TMP']:
        if e in os.environ:
            if '$%s' % e in path:
                path = path.replace('$%s' % e, os.environ[e])
            elif '%%%s%%' % e in path:
                path = path.replace('%%%s%%' % e, os.environ[e])

    if platform.system() == 'Linux':
        path = path.replace('~/', '%s/' % os.environ['HOME']) # ~/x => /home/<user>/x
        path = path.replace('~', '/home/') # ~user/xx => /home/user/xx

    return path


def update_paths(config, analyse, sys, musly, essentia_extractor, bliss_analyser):
    '''
    Update paths settings for detected OS
    '''

    # Set libmusy path, if not already in config
    if musly is not None and not 'lib' in config['musly']:
        _LOGGER.debug('(Defaults) musly.lib set to %s' % musly)
        config['musly']['lib']=musly

    # Update PATH - mainly for windows so that libmusly can find ffmpeg, etc.
    if analyse:
        if sys is not None:
            exe_folder = os.path.join(pathlib.Path(__file__).parent.parent, sys)
            if sys=='windows':
                os.environ['PATH']='%s;%s' % (exe_folder, os.environ['PATH'])
                _LOGGER.debug('Added %s to PATH' % exe_folder)

        # Set essentia extractor path, if not already in config
        if essentia_extractor is not None and not 'extractor' in config['essentia']:
            config['essentia']['extractor']=essentia_extractor
            _LOGGER.debug('(Defaults) essentia.extractor set to %s' % essentia_extractor)
            if not 'highlevel' in config['essentia']:
                config['essentia']['highlevel'] = sys in SUPPORT_ESSENTIA_HIGHLEVEL
                _LOGGER.debug('(Defaults) essentia.highlevel set to %s' % str(config['essentia']['highlevel']))

        # Set bliss analyser path, if not already in config
        if bliss_analyser is not None and not 'analyser' in config['bliss']:
            config['bliss']['analyser']=bliss_analyser
            _LOGGER.debug('(Defaults) bliss.analyser set to %s' % bliss_analyser)

    elif not 'highlevel' in config['essentia']:
        config['essentia']['highlevel'] = sys in SUPPORT_ESSENTIA_HIGHLEVEL
        _LOGGER.debug('(Defaults) essentia.highlevel set to %s' % str(config['essentia']['highlevel']))


def system_name():
    system = platform.system()
    if system == 'Linux':
        return 'linux'
    if system == 'Windows':
        return 'windows'
    if system == 'Darwin':
        return 'mac'
    return 'other'


def setup_paths(config, analyse):
    '''
    Try to configure Musly and Essentia lib/binary paths based upon OS
    '''
    system = system_name()
    arch = platform.architecture()[0]
    proc = platform.processor()
    root_folder = pathlib.Path(__file__).parent.parent

    if system == 'linux':
        if proc == 'x86_64':
            update_paths(config, analyse, 'linux', 'linux/x86-64/libmusly.so', '%s/linux/x86-64/essentia_streaming_extractor_music' % root_folder, '%s/linux/x86-64/bliss-analyse' % root_folder)
            return
        else: # TODO: Check on Pi
            update_paths(config, analyse, 'linux', 'linux/armv7l/libmusly.so', None, None)
            return
    elif system == 'windows':
        if arch.startswith('64'): # 64-bit Python
            update_paths(config, analyse, 'windows', 'windows\\mingw64\\libmusly.dll', '%s\\windows\\streaming_extractor_music.exe' % root_folder, None)
            return
        else:  # 32-bit Python
            update_paths(config, analyse, 'windows', 'windows\\mingw32\\libmusly.dll', '%s\\windows\\streaming_extractor_music.exe' % root_folder, None)
            return
    elif system == 'mac':
        if proc == 'i386':
            update_paths(config, analyse, 'mac', 'mac/i386/libmusly.dylib', '%s/mac/i386/streaming_extractor_music' % root_folder, '%s/mac/i386/bliss-analyse' % root_folder)
            return
    #    else: # M1??? Can use x86_64 binaries on M1?
    #        update_paths(config, analyse, 'mac', 'mac/m1/libmusly.dylib', '%s/mac/streaming_extractor_music' % root_folder, None)
    #        return

    if not 'lib' in config['musly'] or (analyse and config['essentia']['enabled'] and not 'extractor' in config['essentia']):
        exit_with_error('No known binaries, etc, for %s / %s / %s' % (system, proc, arch))


def which(program):
    '''
    Return path of program
    '''
    def is_exe(fpath):
        return os.path.isfile(fpath) and os.access(fpath, os.X_OK)

    fpath, fname = os.path.split(program)
    if fpath:
        if is_exe(program):
            return program
    else:
        for path in os.environ['PATH'].split(os.pathsep):
            exe_file = os.path.join(path, program)
            if is_exe(exe_file):
                return exe_file

    return None


def check_binaries(config):
    '''
    Check that binaries are in PATH
    '''
    apps = []
    if config['essentia']['enabled']:
        apps.append(config['essentia']['extractor'])

    # libmusly uses ffmpeg/ffprobe under Windows
    if system_name()=='windows':
        apps.append('ffmpeg.exe')
        apps.append('ffprobe.exe')

    if len(apps)>0:
        for app in apps:
            appPath = which(app)
            if appPath is None:
                if config['essentia']['enabled'] and app == config['essentia']['extractor']:
                    _LOGGER.info('Essentia extractor not in PATH, Essentia analysis disabled')
                    config['essentia']['enabled'] = False
                else:
                    exit_with_error('%s is not in PATH or is not executable' % app)
            else:
                if app == appPath:
                    _LOGGER.info('Found: %s' % appPath)
                else:
                    _LOGGER.info('Found: %s -> %s' % (app, appPath))


def read_config(path, analyse):
    config={}

    if not os.path.exists(path):
        exit_with_error('%s does not exist' % path)
    try:
        with open(path, 'r') as configFile:
            config = json.load(configFile)
    except ValueError:
        exit_with_error('Failed to parse config file')

    except IOError:
        exit_with_error('Failed to read config file')

    # Check required keys are present
    if not 'paths' in config:
        exit_with_error("'paths' not in config file")

    if analyse:
        for key in ['local', 'db']:
            if not key in config['paths']:
                exit_with_error("'paths.%s' not in config file" % key)
    else:
        if not 'db' in config['paths']:
            exit_with_error("'paths.db' not in config file")

    # Ensure paths end with /
    # and replace HOME, etc, env vars in paths
    for key in config['paths']:
        config['paths'][key] = fix_path(config['paths'][key])
        if not config['paths'][key].endswith(os.sep):
            config['paths'][key]=config['paths'][key]+os.sep

    if 'lmsdb' in config:
        config['lmsdb'] = fix_path(config['lmsdb'])

    # Check paths exist
    if analyse:
        for key in ['local', 'db']:
            if (key=='db' and not os.path.exists(config['paths'][key])) or (analyse and key=='local' and not os.path.exists(config['paths'][key])):
                exit_with_error("'%s' does not exist" % config['paths'][key])
    else:
        if not os.path.exists(config['paths']['db']):
            exit_with_error("'%s' does not exist" % config['paths']['db'])

    if analyse and 'lmsdb' in config:
        if not os.path.exists(config['lmsdb']):
            exit_with_error("'%s' does not exist" % config['lmsdb'])
        if not 'lms' in config['paths']:
            exit_with_error("'paths.lms' not in config file")

    if 'tmp' in config['paths'] and not os.path.exists(config['paths']['tmp']):
        exit_with_error("'%s' does not exist" % config['paths']['tmp'])

    sname = system_name()

    # Check general settings
    if not 'port' in config:
        config['port']=11000

    if not 'host' in config:
        config['host']='0.0.0.0'

    if not 'threads' in config:
        config['threads']=os.cpu_count()

    if not 'simalgo' in config:
        config['simalgo']='bliss' if sname in SUPPORT_BLISS else 'musly'

    if not config['simalgo'] in ['bliss', 'essentia', 'musly', 'mixed', 'simplemixed']:
        exit_with_error("Invalid 'simalgo' setting")

    if not 'bliss' in config:
        config['bliss']={}

    if not 'essentia' in config:
        config['essentia']={}

    if not 'musly' in config:
        config['musly']={}

    setup_paths(config, analyse)

    if not 'enabled' in config['bliss']:
        config['bliss']['enabled'] = (not analyse) or (sname in SUPPORT_BLISS)

    if not 'enabled' in config['essentia']:
        config['essentia']['enabled'] = (not analyse) or ('extractor' in config['essentia'] and os.path.exists(config['essentia']['extractor']))

    if not 'enabled' in config['musly']:
        config['musly']['enabled'] = (not analyse) or (sname not in SUPPORT_BLISS)

    # Check/default musly settings
    if config['musly']['enabled']:
        if not 'lib' in config['musly']:
            exit_with_error("'musly.lib' not in config file" % key)
        else:
            config['musly']['lib'] = fix_path(config['musly']['lib'])
        if not 'extractlen' in config['musly']:
            config['musly']['extractlen']=120
        if not 'extractstart' in config['musly']:
            config['musly']['extractstart']=-210
        if not 'styletracks' in config['musly']:
            config['musly']['styletracks']=1000
        if not 'styletracksmethod' in config['musly']:
            config['musly']['styletracksmethod']='genres'

    # Check/default essentia settings
    if config['essentia']['enabled']:
        if not 'highlevel' in config['essentia']:
            config['essentia']['highlevel']=False
        if not 'filterattrib' in config['essentia']:
            config['essentia']['filterattrib']=True
        if not 'filterattrib_count' in config['essentia']:
            config['essentia']['filterattrib_count']=4
        if not 'filterattrib_lim' in config['essentia']:
            config['essentia']['filterattrib_lim']=0.2
        if not 'filterattrib_cand' in config['essentia']:
            config['essentia']['filterattrib_cand']=0.4
        if not 'attrmix_yes' in config['essentia']:
            config['essentia']['attrmix_yes']=0.6
        if not 'attrmix_no' in config['essentia']:
            config['essentia']['attrmix_no']=0.4

        if analyse:
            if not 'extractor' in config['essentia']:
                exit_with_error("'essentia.extractor' not in config file")
            else:
                config['essentia']['extractor'] = fix_path(config['essentia']['extractor'])
    else:
        config['essentia']['highlevel']=False

    if config['bliss']['enabled']:
        if analyse:
            if not 'analyser' in config['bliss']:
                exit_with_error("'bliss.analyser' not in config file")
            else:
                config['bliss']['analyser'] = fix_path(config['bliss']['analyser'])

    # Check genres, etc.
    if 'genres' in config:
        config['all_genres']=set()
        for i in range(len(config['genres'])):
            config['genres'][i]=set(config['genres'][i])
            config['all_genres'].update(config['genres'][i])

    if 'excludegenres' in config and len(config['excludegenres'])>0:
        config['excludegenres']=set(config['excludegenres'])
    else:
        config['excludegenres']=None

    if 'ignoregenre' in config:
        if isinstance(config['ignoregenre'], list):
            ignore=[]
            for item in config['ignoregenre']:
                ignore.append(tracks_db.normalize_artist(item))
            config['ignoregenre']=set(ignore)
        else:
            config['ignoregenre']=set([config['ignoregenre']])

    if 'normalize' in config:
        tracks_db.set_normalize_options(config['normalize'])

    if not 'minduration' in config:
        config['minduration']=30

    if not 'maxduration' in config:
        config['maxduration']=30*60

    if analyse:
        check_binaries(config)
    elif ('mixed'==config['simalgo'] or 'simplemixed'==config['simalgo']) and not 'mixed' in config:
        exit_with_error('Please deffine algo mix percentages')

    if not config['bliss']['enabled'] and not config['essentia']['enabled'] and not config['musly']['enabled']:
        exit_with_error("Please enable at leasr one of bliss, essentia, musly")

    return config
